fix #9352: instrument change transposition

MarcSabatella opened the pull request:
NOTE: I think this is now ready to merge.

It turns out most of the necessary framework was already in place to get Instrument change text to affect transposition.  After all, we already can retrieving the current instrument for a staff by tick, for playback purposes.  Mostly this PR is just a matter of adding the tick parameter when retrieving the current instrument for transposition purposes as well - and actually, more than a few places we were already doing this (leading to bugs, because it wasn't done consistently).

Basiclly two things left to do:

1) come up with a better solution for Score::transposeKeys() than my current hack of using a zero or octave interval to signify transposition into or out of concert pitch rather than by a fixed interval).  Should be just a matter of adding a nw parameter or two.

2) In my current code, the Staff Properties dialog will show the correct instrument settings according to where you clicked when invoking the dialog.  However, any instrument changes you make end up affecting the entire duration of the score.  We will need to detect the proper range and limit the Score::transpositionChanged() to just that range.  In principle, it shouldn't be hard; I'm just not sure how I want to go about this.

Meanwhile, though, everything else should work well enough to start testing, if anyone is so inclined.

MarcSabatella commented:
Doesn't surprise me that the tests would need to be updated.

Something else that isn't implemented yet: if there are existing notes when you add the Instrument change and/or when you actually change the instrument, they are not automatically transposed.  Similar issues to solve there as with changes to Staff Properties.  But if you set up the Instrument change *first*, it works.

MarcSabatella commented:
Updated - Change Instrument now performs the transposition on the existing notes.  A bug I just discovered that also affects undo prevents the key signature from updating correctly - see https://musescore.org/en/node/78591.  I hope to fix that next, and then think about the backwards compatibility plan (import of scores from 2.0 - 2.0.2).

ericfont commented:
I've tried your branch.  If I add a change instrument from piano to Bb clarinet and place a C#maj keysig at instrument change:
![add c maj](https://cloud.githubusercontent.com/assets/6502474/10008087/f68ff56c-6098-11e5-85a1-ee8c0a462644.png)

and then if I untoggle concert pitch, then the C#maj turns to Ebmaj (I suppose this is desired behavior for readability):
![add c maj then untoggle concert pitch](https://cloud.githubusercontent.com/assets/6502474/10008094/fde5625c-6098-11e5-8061-093c459a3036.png)

But then if the retoggle concert pitch, the Eb changes to Db major:
![add c maj then re-toggle concert pitch](https://cloud.githubusercontent.com/assets/6502474/10008096/0565f794-6099-11e5-8c98-81851fe52ee6.png)

which is enharmonically the same as the original C#maj, but I don't like the fact that it changes from displaying in the C#maj when I first place the C#maj into displaying Dbmaj after unapplying & reapplying concert pitch.  I think the desired behavior should be to either (A) intially start off in C#maj and return to C#maj, or (B) initially start of in Dbmaj and return to Db maj.

Actually, now looking at a fresh score of Bb clarinet only (without any mid-score instrument changes), if I start in concert pitch and apply C#maj, it displays C#maj, but then after untoggle and retoggle concert pitch, I'm in Dbmaj.  So I guess this issue isn't specific to mid-instrument change.   I don't know if this is the most appropriate place to comment, or if I should file a separate issue report (I can't seem to find this in issue tracker, but I might be searching with different words).

MarcSabatella commented:
In order to remember the original key we would need to preserve more info than we do currently - akin to tpc1 and tpc2 for notes.  However, we probably don't need to go that far.  It's a commonly requested feature to add some sort of option to control whether keys are simplified or not when toggling concert pitch.  We should implement that.

MarcSabatella commented:
Now handles transposition when adding or deleting instrument change text as well, and fixes the issue with transposition of key signatures (and also the undo issue).

I think it would also be nice if changing transposition resulted in a key signature being added automatically, but you can also do that manually.

ericfont commented:
I agree would be nice if changing transposition resulted in a key signature being added automatically.

The problem with inserting one manually at the instrument change is if I later decide to change the key signature earlier in the piece, then the manually inserted key sig at the instrument change would not be updated to the new key.

MarcSabatella commented:
There would be no reason automatically-inserted key changes would work any differently from user-inserted ones in this respect.    Either way, changing a key signature earlier in the piece would stop here unless we added some sort of special handling,

ericfont commented:
ok.  I guess I was thinking the automatically added keysignature would be "generated" in which case it would respond automatically to changes in the earlier keysignature.

MarcSabatella commented:
Good point, that's certainly possible.  Not sure if it would really work out that simply, but maybe so.  Definitely worth trying when I have a chance.

MarcSabatella commented:
Turns out it is *not* that simple.  For one thing, setting the key signature to generated means it basically never shows up, because normally, the point of generated key signatures is that they are automatically deleted during layout unless they are at the begining of a system.  Even if I set it to non-generated, things don't work as you'd expect with respect to undo, and even just deleting the instrument change raises a question of whether the key signature should be deleted - what if it was a "real" key signature  that had been placed there manually?  And we'd still be stuck with the basic isse of a non-generated key signature not responding to changes earlier in the score.

So while I'm not saying it's impossible, managing these key signatures automatically is far from easy, and in the end, it might be better not to bother trying.

ericfont commented:
"they are automatically deleted during layout unless they are at the begining of a system"

Well I guess one workaround hack would be to force a new system by inserting a HBox just before :)  (may need to make invisible the clef & timesignature & measure number if not wanted).

MarcSabatella commented:
Of the test failures,  couple I think are expected just need to have the references updated.  But one may be related to my attempt to fix the issue with the undo of the instrument change, and to be honest, I have little confidence in what I did there.  I mean it works for the basic cases I tested, but I wouldn't be at all surprised to find it has side effects.

I think I will back out that aspect of this PR, because I think it needs to be looked at separately.  What it means is that key signatures will not be transposed properly for now with respect to instrument change, but everything else should work. and when the undo bug is fixed, the key signature transposition should fix itself.

I will look at the undo issue later, unless @wschweer or @lasconic wants to.  The cause of the issue is that we are setting the new instrument for the instrument change element *before* we create the undo object here:

https://github.com/musescore/MuseScore/blob/fc1535d7bdbe6bed99c33b957f795f51f518b82e/mscore/propertymenu.cpp#L566-L567

So when ChangeInstrument::flip() tries to preserve the original instrument, it's too late - the new instrument is already there.  So trying to access the old instrument - either to restore it on undo, or to access its transposition info when re-transposing key signatures - gives the wrong result.

I had tried to change this to not set the instrument until within flip(), and was initially confused because there already *is* a call to setInstrument() there.  But there are three different versions of setInstrument() with somewhat different semantics in terms of how deep of a copy is made, and getting it right is tricky.  I sort of suspect we may need to redesign the ChangeInstrument undo object, maybe store an actual Instrument rather than just a pointer, but those kind of changes scare me.

MarcSabatella commented:
Some of the test failures were indeed "real".  These should be fixed by the code I just backed out.  The remaining failing test was expected, and I have updated it.

So as of now, this feels almost ready, it could be merged independently of the fix for the undo issue.  The one thing I definitely still need to do is figure out the strategy for legacy scores.  Because the previous behavior was pretty buggy - different tpc1/tpc2 state depending on whether you added the notes before or after adding the instrument change - it's going to be hard to guarantee scores that use this will come out exactly the same as before.  I'll need to do some more investigation into the previous buggy behaviors to understand what states we might be likely to see scores in.  I'm afraid it's likely to be a mess.  But my goal will be to clean the scores up on import so they are at least internally consistent, and then at most the user might have to manually transpose some passages if we previously relying on some aspect of the buggy behavior that is changed.

MarcSabatella commented:
I think I came up with a pretty good strategy for older scores.  First, I ognire transposition on the instrument change itself - which is to say, I set it the same as the transposition for the staff itself.  Thus, older scores not hit by the bug causing tpc corruption if you add notes after the instrument change will behave just as before.  Then I added code to detect and fix tpc corruption when reading notes.  This same code happens to also fix corruptions introduced by https://musescore.org/en/node/74746, which just coincidentally also just came up today.  Although we will need to fix the underlying bug there separately.  I elected to leave this code for all reads regardless of version, since tpc corruption is bad no matter the cause.

MarcSabatella commented:
I've squashed my commits and rebased (not that much had changed).  I think this is ready to review and merge.

ericfont commented:
I've been using this new instrument change transpose in latest nightly d45cc19 in a sax score that has quite a bit of swapping between bari, alto, tenor, sopr saxes.  I save the score and reopened and did some more editing, but I noticed a strange issue where *some* notes would be displayed exactly an octave higher than normal, both in screen render and pdf export.  I could fix this glitch by right clicking the corresponding change instrument text, selecting change instrument, and re-selecting the instrument that it should be at (indeed they were already previously configured and indicated the correct current instrument) and then when click OK, then all the glitched notes would return to their correct octave.  Even more confusingly, sometimes all notes in a particular section had the glitch, or sometimes it was just a few notes.  

I'm going to look into more and try to find minimum replication steps and create a test score w/ bug report, but too sleepy right now.

MarcSabatella commented:
Just wondering - is the list for 2.1 longer enough that we could consider these (this and #2422, which fixes a regression I introduced here?)  It would be really nice to have instrument changes with transposition working.  As mentioned, things are pretty badly broken in 2.0.x with this respect, so even though the change would mean scores that include such changes would look and/or play differently between 2.0.3 and 2.1, there's no danger of scores not loading or crashing or anything like that, it's not like there is anyone *happy* with how it works in 2.0, and scores not using this feature aren't affected.

I'd  probably need to revisit the compatibility checks I added and submit a separate PR to get this working well for 2.1.  Which I'm totally up for doing, if we agree it is worthwhile.

lasconic commented:
This is very tempting... but if we want 2.1 before the summer, we will need to hurry up. We are on loose schedule right now but I would like to have 2.1 ready before April. So, feel free to try it but hurry up because it will need extensive testing.

MarcSabatella commented:
OK.  I will aim to get this done in next week or so.  I need to review the paper trail and see which connected PR's should go with it.  I remember at one point @wschweer was going to look at the undo issue I mentioned earlier (https://musescore.org/en/node/78591) but I don't know if that ever happened.  It's not a deal breaker, but the side effect of that undo issue was that key signatures wouldn't necessarily transpose properly.
